drivers/i2s/i2s_nrfx.c

```diff
@@ -0,0 +1,708 @@
+		LOG_ERR("Trigger %d not allowed in state %d",
+			cmd, drv_data->state);
+		return -EIO;
+	}
+
+	switch (cmd) {
+	case I2S_TRIGGER_START:
+		drv_data->stop = false;
+		drv_data->active_dir = dir;
+		drv_data->last_tx_buffer = NULL;
+		{
+			int ret = start_transfer(dev, dir);
+
+			if (ret < 0) {
+				return ret;
+			}
+		}
+		drv_data->state = I2S_STATE_RUNNING;
+		break;
+
+	case I2S_TRIGGER_STOP:
+		drv_data->state = I2S_STATE_STOPPING;
+		drv_data->stop = true;
+		break;
+
+	case I2S_TRIGGER_DRAIN:
+		drv_data->state = I2S_STATE_STOPPING;
+		/* If only RX is active, DRAIN is equivalent to STOP. */
+		drv_data->stop = (drv_data->active_dir == I2S_DIR_RX);
+		break;
+
+	case I2S_TRIGGER_DROP:
+		nrfx_i2s_stop();
+		purge_queue(dev, dir);
+		drv_data->state = I2S_STATE_READY;
+		break;
+
+	case I2S_TRIGGER_PREPARE:
+		purge_queue(dev, dir);
+		drv_data->state = I2S_STATE_READY;
+		break;
+
+	default:
+		LOG_ERR("Invalid trigger: %d", cmd);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static const struct i2s_driver_api i2s_nrf_drv_api = {
+	.configure = i2s_nrfx_configure,
+	.config_get = i2s_nrfx_config_get,
+	.read = i2s_nrfx_read,
+	.write = i2s_nrfx_write,
+	.trigger = i2s_nrfx_trigger,
+};
+
+#define I2S(idx) DT_NODELABEL(i2s##idx)
+
+#define I2S_PIN(idx, name)					\
+	COND_CODE_1(DT_NODE_HAS_PROP(I2S(idx), name##_pin),	\
+		    (DT_PROP(I2S(idx), name##_pin)),		\
+		    (NRFX_I2S_PIN_NOT_USED))
+
+#define I2S_NRFX_DEVICE(idx)						     \
+	K_MSGQ_DEFINE(tx_queue##idx, sizeof(void *),			     \
+		CONFIG_I2S_NRFX_TX_BLOCK_COUNT, sizeof(void *));	     \
+	K_MSGQ_DEFINE(rx_queue##idx, sizeof(void *),			     \
+		CONFIG_I2S_NRFX_RX_BLOCK_COUNT, sizeof(void *));	     \
+	static int i2s_nrfx_init##idx(const struct device *dev)		     \
+	{								     \
+		IRQ_CONNECT(DT_IRQN(I2S(idx)), DT_IRQ(I2S(idx), priority),   \
+			    nrfx_isr, nrfx_i2s_irq_handler, 0);		     \
+		irq_enable(DT_IRQN(I2S(idx)));				     \
+		return 0;						     \
+	}								     \
+	static void data_handler##idx(nrfx_i2s_buffers_t const *p_released,  \
+				      uint32_t status)			     \
+	{								     \
+		data_handler(DEVICE_DT_GET(I2S(idx)), p_released, status);   \
+	}								     \
+	static const struct i2s_nrfx_drv_cfg i2s_nrfx_cfg##idx = {	     \
+		.tx_queue = &tx_queue##idx,				     \
+		.rx_queue = &rx_queue##idx,				     \
+		.data_handler = data_handler##idx,			     \
+		.nrfx_def_cfg = NRFX_I2S_DEFAULT_CONFIG(I2S_PIN(idx, sck),   \
+							I2S_PIN(idx, lrck),  \
+							I2S_PIN(idx, mck),   \
+							I2S_PIN(idx, sdout), \
+							I2S_PIN(idx, sdin)), \
+	};								     \
+	static struct i2s_nrfx_drv_data i2s_nrfx_data##idx = {		     \
+		.state = I2S_STATE_READY,				     \
+	};								     \
+	DEVICE_DT_DEFINE(I2S(idx), i2s_nrfx_init##idx, NULL,		     \
+			 &i2s_nrfx_data##idx, &i2s_nrfx_cfg##idx,	     \
+			 POST_KERNEL, CONFIG_I2S_INIT_PRIORITY,		     \
+			 &i2s_nrf_drv_api);
+
+/* Currently, only one I2S instance is supported. */
+I2S_NRFX_DEVICE(0);
```

Comment:
Not blocking, but the above comment might be clearer as:

```
/* Existing SoCs only have one I2S instance */
```

That makes it clear it's not a driver limitation.